operator/sidecar/tbot.go

```diff
@@ -0,0 +1,254 @@
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	"time"
+)
+
+const (
+	DefaultCertificateTTL  = 2 * time.Hour
+	DefaultRenewalInterval = 30 * time.Minute
+)
+
+// ClientAccessor returns a working teleport auth client when invoked.
+// Client users should always call this function on a regular basis to ensure certs are always valid.
+type ClientAccessor func(ctx context.Context) (auth.ClientI, error)
+
+// Bot is a wrapper around an embedded tbot.
+// It implements sigs.k8s.io/controller-runtime/manager.Runnable and
+// sigs.k8s.io/controller-runtime/manager.LeaderElectionRunnable so it can be added to a controllerruntime.Manager.
+type Bot struct {
+	cfg        *config.BotConfig
+	running    bool
+	rootClient auth.ClientI
+	opts       Options
+}
+
+func (b *Bot) initializeConfig() {
+	// Initialize the memory stores. They contain identities renewed by the bot
+	// We're reading certs directly from them
+	rootMemoryStore := &config.DestinationMemory{}
+	destMemoryStore := &config.DestinationMemory{}
+
+	// Initialize tbot config
+	b.cfg = &config.BotConfig{
+		Onboarding: &config.OnboardingConfig{
+			Token:      "",         // Field should be populated later, before running
+			CAPins:     []string{}, // Field should be populated later, before running
+			JoinMethod: types.JoinMethodToken,
+		},
+		Storage: &config.StorageConfig{
+			DestinationMixin: config.DestinationMixin{
+				Memory: rootMemoryStore,
+			},
+		},
+		Destinations: []*config.DestinationConfig{
+			{
+				DestinationMixin: config.DestinationMixin{
+					Memory: destMemoryStore,
+				},
+			},
+		},
+		Debug:           false,
+		AuthServer:      b.opts.Addr,
+		CertificateTTL:  DefaultCertificateTTL,
+		RenewalInterval: DefaultRenewalInterval,
+		Oneshot:         false,
+	}
+
+	_ = b.cfg.CheckAndSetDefaults()
+
+	for _, artifact := range identity.GetArtifacts() {
+		_ = destMemoryStore.Write(artifact.Key, []byte{})
+		_ = rootMemoryStore.Write(artifact.Key, []byte{})
+	}
+
+}
+
+func (b *Bot) GetClient(ctx context.Context) (auth.ClientI, error) {
+	if !b.running {
+		return nil, trace.Errorf("bot not started yet")
+	}
+	// If the bot has not joined the cluster yet or not generated client certs we bail out
+	// This is either temporary or the bot is dead and the manager will shutdown everything.
+	if botCert, err := b.cfg.Storage.Memory.Read(identity.TLSCertKey); err != nil || len(botCert) == 0 {
+		return nil, trace.Retry(err, "bot cert not yet present")
+	}
+	if cert, err := b.cfg.Destinations[0].Memory.Read(identity.TLSCertKey); err != nil || len(cert) == 0 {
+		return nil, trace.Retry(err, "cert not yet present")
+	}
+
+	// Hack to be able to reuse LoadIdentity functions from tbot
+	// LoadIdentity expects to have all the artifacts required for a bot
+	// We loop over missing artifacts and are loading them from the bot storage to the destination
+	for _, artifact := range identity.GetArtifacts() {
```

Comment:
Definitely a bit of a hack. Do you know which bits of information that the destination is missing that prevents it from used to load an identity ? I just partially worry that there's a risk that in future something that belongs to the bot is leaked across to the destination, and this leads to an identity being produced that corresponds to the bot itself rather than the impersonated role.

I'd be curious what we could do (as MachineID) to make your life easier here. One thing that definitely comes to mind, is some kind of channel based events subscription, so that consumers of `tbot` can receive a new client/identity immediately when the tbot refreshes (as part of the cycle, or as part of a CA rotation).

Reply:
>  Do you know which bits of information that the destination is missing that prevents it from used to load an identity ?

I'm still discovering how we create clients and join clusters but from what I understand:
- The destination is missing both TLS and SSH CA certs, which are required to build a working auth client
- Something called by `LoadIdentity` checks the `tokenhash` artifact, as far as I know this is not useful to get a client for the impersonated role

> I'd be curious what we could do (as MachineID) to make your life easier here. One thing that definitely comes to mind, is some kind of channel based events subscription, so that consumers of tbot can receive a new client/identity immediately when the tbot refreshes (as part of the cycle, or as part of a CA rotation).

In my opinion the following changes would make tbot more embeddable:
- as you suggested, a channel to get notified on renewal would be great, here it would allow me to stop creating new clients for every call which is a waste of resources
- a way to init in-memory destinations more easily and safely (it bit me more often that I want to admit). I think generally users should not have to touch the artifacts at all.
- a method to get an `Identity` for a specific bot destination, this would avoid this dirty hack. In its current state `LoadIdentity` needs all bot-only artifacts `sshcacert`, `tlscacert` and `tokenhash`
- a method on `Identity` yielding a client, `auth.ClientI` or maybe `client.Client`

Reply:
Awesome, that seems to match my notes I prepared this morning on what we could do: https://github.com/gravitational/teleport/issues/14507

I'll link your comment here on there as well, thank you.